content/modules/python/pages/writing_functions.adoc

```diff
@@ -0,0 +1,462 @@
+sum_then_multiply_by_x(1,two=2,three=3,x=4) # a positional only argument was passed as a keyword argument, error
+----
+
+== docstrings
+
+docstrings are the strings inside the function immediately following the function declaration. docstrings provide documentation for the function. You can put any information you'd like in a docstring, however, it is best to carefully describe what the function does, and stay consistent in style from docstring to docstring.
+
+You can access a function's docstring in various ways:
+
+[source, Python]
+----
+print(word_count.__doc__)
+----
+
+[source, Python]
+----
+word_count is a function that accepts a sentence as an argument,
+and returns the number of words in the sentence.
+ 
+Args:
+sentence (str): The sentence for which we are counting the words.
+ 
+Returns:
+int: The number of words in the sentence
+ 
+----
+
+[source, Python]
+----
+help(word_count)
+----
+
+[source, Python]
+----
+Help on function word_count in module __main__:
+ 
+word_count(sentence: str) -> int
+word_count is a function that accepts a sentence as an argument,
+and returns the number of words in the sentence.
+     
+Args:
+sentence (str): The sentence for which we are counting the words.
+     
+Returns:
+int: The number of words in the sentence
+----
+
+In addition, if you are coding in a tool like VSCode, for example, you may gain the ability to hover over a function and see its docstring and other information.
+
+It is good practice to write docstrings for every function you write.
+
+== annotations
+
+Another "thing" you may have noticed from our `word_count` function if you've ever used Python in the past. In the _signature_ of our function `def` `word_count(sentence: str) -> int`, we have some extra, _not_ required information in the form of _function annotations_. Specifically, you could write the `word_count` function like this, but it would function just the same:
+
+[source, Python]
+----
+def word_count(sentence):
+    """
+    word_count is a function that accepts a sentence as an argument,
+    and returns the number of words in the sentence.
+
+    Args:
+        sentence (str): The sentence for which we are counting the words.
+
+    Returns:
+        int: The number of words in the sentence
+    """
+    
+    def _strip_punctuation(sentence: str):
+        """
+        helper function to strip punctuation.
+        """
+        return sentence.translate(str.maketrans('', '', string.punctuation))
+    
+    sentence_no_punc = _strip_punctuation(sentence)
+    result = len(sentence_no_punc.split())
+    return result
+----
+
+Here, we do not specify that `sentence` should be a `str` or that the returned result should be an `int`. When we _do_ annotate functions, it is purely a way to add metadata to our function. In large projects, function annotations are recommended. Although Python does not strictly enforce type annotations, packages like `mypy` can be added to a deployment scheme to strictly enforce it.
```

Comment:
In the example above, the expected types for `sentence` and output are removed. Annotating a function is a way to add data about the function (metadata) to the function. Type annotations are strongly encouraged in large or collaborated projects. 

NOTE: Packages like `mypy` will enforce type annotations. They can be tools to help you develop good code practices.